src/dune/coq_lib.ml

```diff
@@ -0,0 +1,152 @@
+  let make ?loc ?hints paragraphs =
+    Error (User_error.E (User_error.make ?loc ?hints paragraphs))
+
+  let duplicate_theory_name theory =
+    let loc, name = theory.Dune_file.Coq.name in
+    let name = Coq_lib_name.to_string name in
+    make ~loc [ Pp.textf "Duplicate theory name: %s" name ]
+
+  let theory_not_found ~loc name =
+    let name = Coq_lib_name.to_string name in
+    make ~loc [ Pp.textf "Theory %s not found" name ]
+
+  let private_deps_not_allowed ~loc private_dep =
+    let name = Coq_lib_name.to_string (name private_dep) in
+    make ~loc
+      [ Pp.textf
+          "Theory %S is private, it cannot be a dependency of a public theory. \
+           You need to associate %S to a package."
+          name name
+      ]
+
+  let duplicate_boot_lib ~loc boot_theory =
+    let name = Coq_lib_name.to_string (snd boot_theory.Dune_file.Coq.name) in
+    make ~loc [ Pp.textf "Cannot have more than one boot library: %s)" name ]
+
+  let cycle_found ~loc cycle =
+    make ~loc
+      [ Pp.textf "Cycle found"
+      ; Pp.enumerate cycle ~f:(fun t ->
+            Pp.text (Coq_lib_name.to_string (snd t.name)))
+      ]
+end
+
+module DB = struct
+  type lib = t
+
+  type nonrec t =
+    { boot : (Loc.t * t) option
+    ; libs : t Coq_lib_name.Map.t
+    }
+
+  let boot_library { boot; _ } = boot
+
+  let create_from_stanza ((dir, s) : Path.Build.t * Dune_file.Coq.t) =
+    let name = snd s.name in
+    ( name
+    , { name = s.name
+      ; wrapper = Coq_lib_name.wrapper name
+      ; obj_root = dir
+      ; src_root = dir
+      ; theories = s.theories
+      ; libraries = s.libraries
+      ; package = s.package
+      } )
+
+  (* Should we register errors and printers, or raise is OK? *)
+  let create_from_coqlib_stanzas sl =
+    let libs =
+      match Coq_lib_name.Map.of_list_map ~f:create_from_stanza sl with
+      | Ok m -> m
+      | Error (_name, _w1, (_, w2)) ->
+        Result.ok_exn (Error.duplicate_theory_name w2)
+    in
+    let boot =
+      match List.find_all ~f:(fun (_, s) -> s.Dune_file.Coq.boot) sl with
+      | [] -> None
+      | [ l ] -> Some ((snd l).loc, snd (create_from_stanza l))
+      | _ :: (_, s2) :: _ ->
+        Result.ok_exn (Error.duplicate_boot_lib ~loc:s2.loc s2)
+    in
+    { boot; libs }
+
+  let resolve ?(allow_private_deps = true) db (loc, name) =
+    match Coq_lib_name.Map.find db.libs name with
+    | Some s ->
+      if (not allow_private_deps) && Option.is_none s.package then
+        Error.private_deps_not_allowed ~loc s
+      else
+        Ok s
+    | None -> Error.theory_not_found ~loc name
+
+  let find_many t ~loc = Result.List.map ~f:(fun name -> resolve t (loc, name))
+
+  (* Where should we move this? *)
+  module Result_monad : Monad_intf.S with type 'a t = ('a, exn) Result.t =
+  struct
+    type 'a t = ('a, exn) Result.t
+
+    let return x = Ok x
+
+    let ( >>= ) = Result.O.( >>= )
+  end
+
+  module Coq_lib_closure = Top_closure.Make (String.Set) (Result_monad)
```

Comment:
Why don't you use `Result`?

Reply:
You mean in place of `Result_monad`? I didn't find a place with the right signature; I could adapt the code tho to add this to stdune but I thought we'd discuss it first.

Reply:
Yeah, agreed. I ended up replacing it with `Or_exn` by adding the appropriate functions to `Or_exn`.